doc/man/man4/cons.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/cons.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH CONS 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B cons, s_cons
\- Connection-Oriented Network Service (CONS) Conversion Module
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <unistd.h>
.br
.B #include <sys/stropts.h>
.br
.B #include <sys/cons.h>
.HP
.BI "int " x25_stream " = open(" x25_device ", " flags );
.PD 0
.HP
.BI "int " retval " = ioctl(" x25_stream ", " \fRI_PUSH\fP ", " cons_module );
.PD
.HP
.IB cons_stream " = open(" \(lq/dev/cons\(rq ", O_RDWR);
.PD 0
.HP
.IB cons_stream " = open(" \(lq/dev/cons_cm\(rq ", O_RDWR);
.PD
.\"
.\"
.SH DESCRIPTION
.PP
.B CONS
is an implementation of X.213
.[
x213
.]
Connection-oriented Network Service (CONS).
.PP
There are several ways of providing the ISO CONS as follows:
.IP 1. \w'0.\(em'u
The classical way is using
X.25 as described in X.223.
.[
x223
.]
This is suitable both for X.25/LAPB as described in X.223
.[
x223
.]
as well as X.25/LLC2 as described in ISO/IEC 8881.
.[
ISO/IEC 8881
.]
This approach can utilize a SNARE
.[
ISO/IEC 10030
.]
or ARE
.[
x116
.]
for both address and route determination
using configuration information or redirection information.
.IP 2.
Another way is providing CONS over X.25, but providing X.25 using an X.25
over TCP protocol such as
.BR xot (4)
as described in RFC 1613,
.[
rfc1613
.]
or a similar SCTP protocol,
.BR xos (4).
These are not symmetrical protocols in the sense that one XOT/XOS system
accesses X.25 links available on the other, while the other serves X.25 links
to XOT/XOS systems.  As such it is a client-server protocol.  It is not very
useful to have systems act in both roles.  Because XOT/XOS systems of the
server variety are normally routers (and typically Cisco routers), this
implementation, at least initially, only implements the network user role.
.IP 3.
Another way of providing ISO CONS is to use a CONS over TCP/IP protocol,
.BR isot (4),
as described in RFC 1006,
.[
rfc1006
.]
or a similar SCTP protocol,
.BR isos (4).
This is only suitable for use by TP0.
This is a symmetrical peer-to-peer protocol.
.IP 4.
Another CONS over TCP/IP protocol,
.BR itot (4),
is described in RFC 2126,
.[
rfc2126
.]
or a similar SCTP protocol,
.BR itos (4).
This approach is suitable for use by TP0 or TP2 (and TP4 for SCTP).
This is a symmetrical peer-to-peer protocol.
.\"
.\"
.SH USAGE
.PP
A user of the CONS service opens the pseudo-device driver by opening a the
clone device
.RB /dev/cons .
This yields a CONS Stream.  When the user binds the Stream to a local NSAP,
the form of the NSAP determines which of the underlying protocols are used to
provide service to the CONS Stream.
.PP
A CONS Stream is associated with an underlying provider in the following
fashion:
.IP \(bu \w'\(bu\(em'u
When the CONS Stream is bound, using the
.BR N_BIND_REQ (7)
primitive, the driver uses the NSAP to determine whether an existing provider
exists, and if it does, the Stream is associated with the provider.  If a
suitable provider does not yet exist, a request is issued to a connection
manager daemon listening on the connection management Stream.
.IP \(bu
The connection manager satisfies requests by examining the NSAP contained in
the
.BR N_BIND_REQ (7),
primitive.  The NSAP can be one of 2 forms:
.sp
.RS
.IP 1. \w'0.\(em'u
An X.121 address.
.IP ""
In this case the connection manager opens an X.25 PLP
stream on the X.25 PLP device, binds the Stream appropriately according to the
information in the
.BR N_BIND_REQ (7)
primitive, and then links the bound Stream under the CONS multiplexing driver.
It responds to the request indicating the multiplexer index of the linked
Stream.
.IP ""
If binding the X.25 PLP Stream failes, the connection manager responds on the
connection management Stream with a negative acknowledgement that the CONS
driver translates into an
.BR N_ERROR_ACK (7)
primitive to the CONS user.
.IP 2.
An RFC 1277,
.[
1277
.]
RFC 1888,
.[
rfc1888
.]
or
RFC 4548
.[
rfc4548
.]
address.
.IP ""
In this case the connection manager opens an SCTP, TCP or UDP Stream and pushes
the appropriate
.BR isot "(4), " itot "(4), " isos "(4) or " isot (4)
module on the Stream.
The Stream is then bound in the same fashion as an X.25 PLP Stream using an
.BR N_BIND_REQ (7)
primitive.  If the Stream is bound successfully, the Stream is linked under
the CONS multiplexing driver and the connection manager responds on the
connection management Stream indicating the multiplexer index of the linked
Stream.
.IP ""
If binding the IXOX Stream fails, the connection manager response on the
connection managment Stream with a negative acknowledgement that the CONS
driver translates into an
.BR N_ERROR_ACK (7)
primitive to the CONS user.
.RE
.IP \(bu \w'\(bu\(em'u
Provider Streams may also be linked under the CONS multiplexing driver at
configuration time using static configuration information obtained from
configuration files.
.IP \(bu
Provider Streams may also be linked under the CONS multiplexing driver using
route discovery protocols.
.PP
Note that the CONS driver does not perform X.25 nor IP routing.  It only
matches CONS user Stream from the upper multiplex to provider Streams on the
lower multiplex.  X.25 routing is performed within the X.25 PLP driver.  IXOX
routing is performed by the TCP/IP subsystem.
.PP
Note also, that when the CONS connection manager opens an X.25 PLP Stream and
binds it, that it may subsequently invoke an X.25 connection manager to
satisfy the request in a similar fashion, but for the X.25 PLP multiplexing
driver.
.\"
.\"
.SH INTERFACE
.PP
The CONS user interface is the Network Provider Interface (NPI),
.[
npi revision 2.0.0 draft 2
.]
.BR npi (7),
supporting the CONS service interface.
.[
x213
.]
.PP
The CONS connection manager interface is a specialized interface specific to
the CONS driver.
.\"
.\"
.SH ADDRESSES
.PP
CONS supports X.25 addresses as defined in X.121,
.[
x121
.]
CONS NSAPs as defined in X.213,
.[
x213
.]
and Internet NSAPs as defined in RFC 1277,
.[
rfc1277
.]
RFC 1888,
.[
rfc1888
.]
and RFC 4548.
.[
rfc4548
.]
.\"
.\"
.SH OPTIONS
.PP
CONS user Stream support the full set of Quality of Service parameters as
described in X.213;
.[
x213
.]
however, negotiation of the Qualitfy of Service parameters is performed
between the CONS user and the underlying CONS provider.  Therefore, QOS
parameters supported to the CONS user is dependent upon the QOS support of the
underlying provider.
.PP
For example, QOS parameters included in a
.BR N_CONN_REQ "(7), " N_CONN_RES "(7) or " N_OPTMGMT_REQ (7)
primitive are passed to the equivalent primitives to the underlying
provider.  Negotiation responses come directly from the underlying provider.
.\"
.\"
.SH PRIMITIVES
.PP
Primitives supported by CONS user Streams are the fully set of OSI conforming
Network Provider Interface (NPI),
.[
npi revision 2.0.0 draft 2
.]
.BR npi (7),
primitives for Connection-oriented Network Service.
.PP
CONS connection management Streams support a specialized set of primitives,
based on the NPI primitives, that provide the ability for the connection
manager to receive requests for new connections as well as the ability to
indicate the result of those requests.
.PP
CONS CM requests and responses are formatted as follows:
.PP
.nf
\fC\s-1\
struct cons_mgmt {
    np_ulong cm_user;           /* User index */
    np_ulong cm_prov;           /* Provider index */
    union N_primitives cm_prim; /* Primitive */
};
\s+1\fP
.fi
.TP \w'\fIcm_user\fR\(em'u
.I cm_user
is the user index.
.TP
.I cm_prov
is the provider index.
.TP
.I cm_prim
is the primitive.  The exact format of the primitive and the size of the
remainder of the structure can be determined by the
.I cm_prim.type
field of the
.B N_primitives
union.  The
.B N_primitives
union is defined in the
.RB < sys/npi.h >
header file.
.PP
The
.B cons_mgmt
structure is contained in the control part
.RB ( M_PROTO "(9) or " M_PCPROTO (9)
message block) of the request or response.
Any data is contained in the data part
.RB ( M_DATA (9)
message blocks) of the request or response.  The type of the command is
determined from the
.I cm_prim.type
field in the
.B cons_mgmt
structure.
The
.B cons_mgmt
structure is defined in the
.RB < sys/cons.h >
header file.
.PP
Overloaded primitives are as follows:
.TP
.B Bind Request
The bind request is issued to the CONS CM on the connection management Stream
when a suitable provider could not be found to satisfy a CONS User bind
request,
.BR N_BIND_REQ (7).
This primitive requires an acknowledgement by the connection manager with
either a bind acknowledgement or an error acknowledgement.
.RS
.TP \w'\fIcm_user\fR\(em'u
.I cm_user
Indicates the index within the CONS driver of the CONS User at which the
request originated.  This is an opaque identifier that the CONS CM uses in the
bind or error acknowledgement to identify the transaction.  The user index is
most easily implemented as the device number of the CONS User Stream.
.TP
.I cm_prov
Ignored in the bind request.
.TP
.I cm_prim
Contains the
.BR N_BIND_REQ (7)
primitive as it was issued by the CONS User.
.RE
.TP
.B Bind Acknowledgement
The bind acknowledgement is issued by the CONS CM on the connection management
Stream when a suitable provider has been bound and linked beneath the CONS
multiplexing driver.
.RS
.TP \w'\fIcm_user\fR\(em'u
.I cm_user
Specifies the user index from the corresponding bind request.  The user index
is most easily implemented as the device number of the CONS User Stream.
.TP
.I cm_prov
Specifies the multiplexer identifier (returned by the
.BR I_LINK "(7) or " I_PLINK (7)
operation) that identifies the provider Stream to which the bind ocurred.
.TP
.I cm_prim
Contains the
.BR N_BIND_ACK (7)
primitive as it should be issued to the CONS User.
.RE
.TP
.B Connection Request
The connection request is issued to the CONS CM on the connection mangement
Stream when a suitable provider could not be found to satify a CONS User
connection request,
.BR N_CONN_REQ (7).
This primitive requires an acknowledgement by the connection manager with
either a connection confirmation, disconnect indication, or error
acknowledgement.
.RS
.TP \w'\fIcm_user\fR\(em'u
.I cm_user
Indicates the user index from the corresponding connection request.  The user
index is most easily implemented as the device number of the CONS User Stream.
.TP
.I cm_prov
Indicates the provider index from the corresponding connection request.  The
provider index is implemented as the multiplexer identifier of the CONS
provider Stream.  This is the provider that was associated with the CONS user
during the bind request.
.TP
.I cm_prim
Contains the
.BR N_CONN_REQ (7)
primitive as it was issued by the CONS User.  Any data associated with the
connection request is provided in
.BR M_DATA (9)
message blocks attached to the primitive.
.RE
.TP
.B Connection Confirmation
The connection confirmation is issued by the CONS CM on the connection
management Stream when a suitable provider has been connected and linked
beneath the CONS multiplexing driver.
.RS
.TP \w'\fIcm_user\fR\(em'u
.I cm_user
Specifies the user index from the corresponding connection request.
.TP
.I cm_prov
Specifies the multiplexer identifier (returned by the
.BR I_LINK "(7) or " I_PLINK (7)
operation) that identifies the provider Stream on which the connection
ocurred.
.TP
.I cm_prim
Contains the
.BR N_CONN_CON (7)
primitive as it should be issued ot the CONS User.  Any data associated with
the connection request is provided in the
.BR M_DATA (9)
message blocks attached to the primitive.
.RE
.TP
.B Disconnect Indication
The disconnect indication is issued by the CONS CM on the connection
management Stream when a connection to the CONS provider was not possible and
resulted in a disconnect.
.RS
.TP \w'\fIcm_user\fR\(em'u
.I cm_user
Specifies the user index from the corresponding connection request.
.TP
.I cm_prov
Specifies the provider index from the corresponding connection request.
.TP
.I cm_prim
Contains the
.BR N_DISCON_IND (7)
primitive as it should be issued to the CONS User.
.RE
.TP
.B Error Acknowledgement
The error acknowledgement is issued by the CONS CM on the connection
management Stream when the bind or connection to the CONS provider was not
possible and resulted in an error.
.RS
.TP \w'\fIcm_user\fR\(em'u
.I cm_user
Specifies the user index from the corresponding bind or connection request.
.TP
.I cm_prov
Specifies the provider index from the corresponding bind or connection
request.
.TP
.I cm_prim
Contains the
.B N_ERROR_ACK (7)
primitive as it should be issued to the CONS User.
.RE
.\"
.\"
.SH IOCTLS
.PP
CONS Streams do not support any CONS-specific input-output controls.  The CONS
connection management Stream supports the
.BR I_LINK "(7), " I_PLINK "(7), " I_UNLINK "(7), " I_PUNLINK (7)
.I STREAMS
input-output controls described under
.BR streamio (7).
When a CONS CM links a provider Stream under the mulitiplexing driver using
.BR I_LINK "(7) or " I_PLINK (7),
the CONS driver determines the characteristics of the provider by issuing a
.BR N_INFO_REQ (7)
primitive on the provider Stream and processing the
.BR N_INFO_ACK (7)
response.
.\"
.\"
.SH NOTICES
.PP
One of the primary purposes of the CONS driver was so that the ISO Transport
driver,
.BR tp (4),
does not have to perform specialized connection management.  Unfortunately
this is not the case.
.BR isot "(4), " itot "(4), " isos "(4) and " isot (4)
cannot be used as general purpose CONS replacements.  They can only be used
for specific ISO Transport Protocol Classes and can only be used to support
ISO Transport.  Therefore, this CONS driver is not implemented as described
here.  This information is historical.
.\"
.\"
.SH IMPLEMENTATION
.PP
Not implemented.
.\"
.\"
.SH FILES
.PP
.RB < sys/cons.h >
The connection manager header file containing the definition of the
.B cons_mgmt
structure.
.PP
The following
.BR /etc/netconfig
entries and Name-to-Address translation libraries are used with the CONS driver
in support of TI-RPC:
.sp
.RS
.nf
\fC\s-1\
clnp    tpi_clts    v   osi     datagram_v  /dev/clnp   osiaddr.so
\s+1\fP
.fi
.RE
.TP
.B /usr/lib/osiaddr.so
Provides the Name-to-Address translation library for OSI.  See
.BR osiaddr (5).
.PP
The following
.B /etc/sock2path
entries are used with the CONS driver in support of the Sockets API:
.sp
.RS
.nf
\fC\s-1\
PF_OSI  SOCK_DGRAM      0               /dev/clnp
PF_OSI  SOCK_DGRAM      OSIPROTO_CLNS   /dev/clnp
\s+1\fP
.fi
.RE
.\"
.\"
.SH MODULES
.PP
Modules most often used with the CONS driver are as follows:
.TP
.RI \(lq isot \(rq
This is the
.BR isot (4)
module that transforms a
.BR tcp "(4) or " udp (4)
Stream into a CONS provider Stream supporting ISO Transport Protocol Class 0
(TP0).
.TP
.RI \(lq itot \(rq
This is the
.BR itot (4)
module that transforms a
.BR tcp "(4) or " udp (4)
Stream into a CONS provider Stream supporting ISO Transport Protocol Class 0
(TP0) or Class 2 (TP2).
.TP
.RI \(lq isos \(rq
This is the
.BR isos (4)
module that transforms a
.BR sctp (4)
Stream into a CONS provider Stream supporting ISO Transport Protocol Class 0
(TP0).
.TP
.RI \(lq itos \(rq
This is the
.BR itos (4)
module that transforms a
.BR sctp (4)
Stream into a CONS provider Stream supporting ISO Transport Protocol Class 0
(TP0), Class 2 (TP2) or Class 4 (TP4).
.\"
.\"
.SH DEVICES
.PP
The CONS driver provides two devices:
.TP
.B /dev/cons
This is a clone device that provides user CONS Streams.  Each open of this
device provides a new CONS user Stream that begins life in the unbound state.
.TP
.B /dev/cons_cm
This is a clone device that provides connection management Streams.  Each open
of this device provides a new Connection Management Stream.  Multiple
connection management streams can be opened.
.\"
.\"
.SH SEE ALSO
.PP
.IR ISO (4).
.\"
.\"
.SH BUGS
.PP
Because the CONS driver is not implemented, it has no bugs.  Or, all the bugs
were in the design.
.\"
.\"
.SH COMPATIBILITY
.PP
None.
.\"
.\"
.SH CONFORMANCE
.PP
None.
.\"
.\"
.SH HISTORY
.PP
None.
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/drivers/cons.c
/*****************************************************************************

 @(#) File: src/drivers/cons.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/cons.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  This is an X.223 CONS driver.
 *
 *  The driver implements X.223 (support of CONS using X.25) as a multiplexing pseudo-device driver.
 *  Per X.223, it implements the X.25 PLP in support of the CONS service interface.  The driver
 *  presents a NPI CONS service interface at the upper multiplex.
 *
 *  Data links Streams providing the data link service using the DLPI interface are linked beneath
 *  the multiplexing driver.  Data link Streams should be attached, bound and connected.  Data link
 *  Streams can be Streams of any type.  In support of X.223 they can be LLC2 connections per
 *  ISO/IEC 8881, ISDN connections per X.612 ISO/IEC 9574, X.21/X.21 bis PADS or X.30 ISDN TAs per
 *  X.613 ISO/IEC 10588, or PSTN per X.614 ISO/IEC 10732.
 *
 *  Routing:
 *
 *  X.121 address-based routing is performed by associating a network title (IS) or address (ES)
 *  with each data link to form a routing table.  Network titles or addresses are associated with a
 *  data link once it has been linked beneath the multiplexing driver using input-output controls.
 *  Packets received that are associated with external addresses are routed out another data link or
 *  rejected.  Packets received that are associated with a locally assigned address are processed
 *  against the CONS interface.
 *
 */

/*
 * vim: tw=100
 */
